# Exploring Linear Algebra Using Python

This notebook uses Python and NumPy to explore several basic ideas from linear algebra. The examples cover vector calculations, matrix operations, and eigenvalue-related computations.



## 1. Learning Goals
By completing this notebook, I will:

* Work with vectors and perform basic vector calculations.
* Demonstrate multiplication of a vector by a scalar.
* Find dot products and vector magnitudes.
* Perform addition and multiplication with matrices.
* Find matrix transposes, inverses, and ranks.
* Calculate eigenvalues and eigenvectors.
* Explain the meaning of the numerical results produced by NumPy.



## 2. Importing the Required Library

To work with vectors, matrices, and linear algebra functions in this notebook, we import the NumPy library.


In [2]:
import numpy as np

## 3. Vector Calculations

Vectors are one-dimensional arrays of numbers that represent quantities such as direction, movement, or data points. In linear algebra, vectors are used to model many real-world relationships, from coordinates in space to features in a dataset. Python and NumPy make it easy to perform vector-based calculations efficiently.


### 3.1 Adding Vectors

Vector addition is done by adding corresponding elements from two vectors. This operation combines two vectors while preserving their direction and magnitude in a structured way.

In [4]:
vector_a = np.array([1, 2, 3])
vector_b = np.array([4, 5, 6])

vector_sum = vector_a + vector_b

print("Vector A:", vector_a)
print("Vector B:", vector_b)
print("Vector Sum:", vector_sum)

Vector A: [1 2 3]
Vector B: [4 5 6]
Vector Sum: [5 7 9]


**Interpretation:** The corresponding elements of the two vectors are added together. Therefore, the sum of `[1, 2, 3]` and `[4, 5, 6]` is `[5, 7, 9]`.

### 3.2 Multiplying a Vector by a Scalar


Scalar multiplication means multiplying each element of a vector by the same number. This changes the vector’s size without changing its direction, unless the scalar is negative, in which case the direction is reversed.


In [5]:
vector = np.array([1, 2, 3])
scalar = 3

result = scalar * vector

print("Vector:", vector)
print("Scalar:", scalar)
print("Result:", result)

Vector: [1 2 3]
Scalar: 3
Result: [3 6 9]


**Interpretation:** Each element of the vector was multiplied by the scalar value 3. Therefore, `[1, 2, 3]` becomes `[3, 6, 9]`.

### 3.3 Dot Product and Vector Magnitude

The dot product measures how closely two vectors align with each other by multiplying corresponding elements and summing the results. The norm of a vector represents its length or magnitude, which is useful in many mathematical and machine learning applications.

In [6]:
vector_a = np.array([1, 2, 3])
vector_b = np.array([4, 5, 6])

dot_product = np.dot(vector_a, vector_b)
norm_a = np.linalg.norm(vector_a)

print("Vector A:", vector_a)
print("Vector B:", vector_b)
print("Dot Product:", dot_product)
print("Norm of Vector A:", norm_a)

Vector A: [1 2 3]
Vector B: [4 5 6]
Dot Product: 32
Norm of Vector A: 3.7416573867739413


**Interpretation:** The dot product of the two vectors is 32. The norm of Vector A is approximately 3.74, which represents its magnitude.

## 4. Matrix Operations

A matrix is a rectangular arrangement of numbers in rows and columns. Matrices are fundamental in linear algebra because they can represent systems of equations, datasets, transformations, and many other structures in data science and engineering.


### 4.1 Adding and Multiplying Matrices

Matrix addition is performed element by element, while matrix multiplication follows a different rule that combines rows and columns. Matrix multiplication is especially important because it allows one matrix to transform or interact with another in a meaningful way.


In [9]:
matrix_a = np.array([
    [1, 2],
    [3, 4]
])

matrix_b = np.array([
    [5, 6],
    [7, 8]
])

matrix_sum = matrix_a + matrix_b
matrix_product = matrix_a @ matrix_b

print("Matrix A:")
print(matrix_a)

print("\nMatrix B:")
print(matrix_b)

print("\nMatrix Addition:")
print(matrix_sum)

print("\nMatrix Multiplication:")
print(matrix_product)

Matrix A:
[[1 2]
 [3 4]]

Matrix B:
[[5 6]
 [7 8]]

Matrix Addition:
[[ 6  8]
 [10 12]]

Matrix Multiplication:
[[19 22]
 [43 50]]


### 4.2 Transpose and Inverse of a Matrix

The transpose of a matrix swaps its rows and columns. The inverse of a square matrix is another matrix that, when multiplied with the original, produces the identity matrix. Inverse matrices are useful for solving systems of equations.


In [10]:
matrix = np.array([
    [1, 2],
    [3, 4]
])

transpose = matrix.T
inverse = np.linalg.inv(matrix)

print("Original Matrix:")
print(matrix)

print("\nTranspose:")
print(transpose)

print("\nInverse:")
print(inverse)

Original Matrix:
[[1 2]
 [3 4]]

Transpose:
[[1 3]
 [2 4]]

Inverse:
[[-2.   1. ]
 [ 1.5 -0.5]]


### 4.3 Matrix Rank

The rank of a matrix indicates the number of linearly independent rows or columns it contains. This measure helps determine whether a matrix has full rank and whether a system of equations has a unique solution.

In [11]:
matrix = np.array([
    [1, 2],
    [3, 4]
])

rank = np.linalg.matrix_rank(matrix)

print("Matrix:")
print(matrix)

print("\nRank:", rank)

Matrix:
[[1 2]
 [3 4]]

Rank: 2


## 5. Eigenvalues and Eigenvectors
Eigenvalues and eigenvectors are important concepts in linear algebra that describe how a matrix transforms space. For a given square matrix, an eigenvector is a non-zero vector whose direction remains unchanged after multiplication by the matrix, while the eigenvalue tells how much the vector is scaled.


### 5.1 Understanding the Concept

When a matrix acts on an eigenvector, the output is simply the same vector multiplied by a scalar value. This property makes eigenvalues and eigenvectors especially useful in areas such as physics, computer graphics, and machine learning.

### 5.2 Computing in Python

Using NumPy, we can efficiently compute eigenvalues and eigenvectors for a square matrix. This is useful when analyzing the behavior of linear transformations and understanding the structure of a matrix.


In [12]:
matrix = np.array([
    [2, 0],
    [0, 3]
])

eigenvalues, eigenvectors = np.linalg.eig(matrix)

print("Matrix:")
print(matrix)

print("\nEigenvalues:")
print(eigenvalues)

print("\nEigenvectors:")
print(eigenvectors)

Matrix:
[[2 0]
 [0 3]]

Eigenvalues:
[2. 3.]

Eigenvectors:
[[1. 0.]
 [0. 1.]]


### 5.3 Checking and Interpreting the Results

The eigenvalues and eigenvectors returned by NumPy provide information about how the matrix stretches or compresses space. The eigenvalues represent the scale factors, and the eigenvectors show the directions that are preserved under the transformation.

## 6. Final Remarks

This notebook explored several key ideas in linear algebra using Python and NumPy, including vector operations, matrix manipulation, and eigenvalue analysis. These topics form the foundation of many computational methods used in science, engineering, and artificial intelligence.

The examples show how the same mathematical principles can be translated into code and used to solve practical problems. By combining theory with numerical computation, we can better understand how linear algebra supports modern data analysis and machine learning.